stackslib/src/net/api/getclaritymarfvalue.rs

```diff
@@ -0,0 +1,218 @@
+// Copyright (C) 2013-2020 Blockstack PBC, a public benefit corporation
+// Copyright (C) 2020-2024 Stacks Open Internet Foundation
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use clarity::vm::clarity::ClarityConnection;
+use clarity::vm::representations::{
+    MARF_KEY_FOR_QUAD_REGEX_STRING, MARF_KEY_FOR_TRIP_REGEX_STRING,
+};
+use regex::{Captures, Regex};
+use stacks_common::types::net::PeerHost;
+use stacks_common::util::hash::to_hex;
+
+use crate::net::http::{
+    parse_json, Error, HttpNotFound, HttpRequest, HttpRequestContents, HttpRequestPreamble,
+    HttpResponse, HttpResponseContents, HttpResponsePayload, HttpResponsePreamble,
+};
+use crate::net::httpcore::{
+    request, HttpPreambleExtensions, HttpRequestContentsExtensions, RPCRequestHandler,
+    StacksHttpRequest, StacksHttpResponse,
+};
+use crate::net::{Error as NetError, StacksNodeState, TipRequest};
+
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct ClarityMarfValueResponse {
+    pub data: String,
+    #[serde(rename = "proof")]
+    #[serde(default)]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub marf_proof: Option<String>,
+}
+
+#[derive(Clone)]
+pub struct RPCGetClarityMarfValueRequestHandler {
+    pub clarity_marf_key: Option<String>,
+}
+impl RPCGetClarityMarfValueRequestHandler {
+    pub fn new() -> Self {
+        Self {
+            clarity_marf_key: None,
+        }
+    }
+}
+
+/// Decode the HTTP request
+impl HttpRequest for RPCGetClarityMarfValueRequestHandler {
+    fn verb(&self) -> &'static str {
+        "GET"
+    }
+
+    fn path_regex(&self) -> Regex {
+        Regex::new(&format!(
+            r"^/v2/clarity_marf_value/(?P<clarity_marf_key>(vm-epoch::epoch-version)|({})|({}))$",
```

Comment:
I strongly disagree with using the MARF key as input.  To do so would require an explanation to an RPC API consumer of how to construct a Clarity MARF key, which essentially requires them to read the majority of [SIP-005](https://github.com/stacksgov/sips/blob/main/sips/sip-005/sip-005-blocks-and-transactions.md).  Instead, you should do the following:

* Accept the _hash_ of the key as input.  Then, you don't have to worry about verifying that it is well-formed, and you don't have to worry about explaining to the RPC API consumer how to construct it.  Furthermore, it's much safer and easier to determine if a string is a well-formed hash instead of a well-formed Clarity MARF key.
* Update the Clarity DB to take the hash as an argument, instead of the key (in support of the above).

Reply:
Also, I dislike the path you chose, since it doesn't leave room for semantic expansion.  Instead, can you use the following:  `/v2/marf/clarity/{key-hash}`, where `key-hash` is the hash of the SIP-005 key.

Reply:
In my use case (of Clarinet consuming this API), I already know the key (but not the hash). But I agree that it's not ideal to have the key as a use input.
Can you guide me to some code where I could see how to get the key <> key_hash conversion?

> Update the Clarity DB to take the hash as an argument, instead of the key (in support of the above).

Not sure to follow here, where should I perform this change?

--- 

I think I just lack knowledge about this repo to fully address this comment, thanks for providing extra context 🙏

Reply:
Sure -- the (string) key's bytes are hashed via SHA512/256 to produce the MARF key.

> Not sure to follow here, where should I perform this change?

The trait `ClarityBackingStore` needs to be modified to add a method like `get_data_with_proof()`, but which takes an implementation of the `MarfTrieId` trait instead of a `&str` key.  There currently isn't a way to query the MARF via the ClarityDB with a bare hash.

Reply:
> I strongly disagree with using the MARF key as input. To do so would require an explanation to an RPC API consumer of how to construct a Clarity MARF key, which essentially requires them to read the majority of [SIP-005](https://github.com/stacksgov/sips/blob/main/sips/sip-005/sip-005-blocks-and-transactions.md).

Why would getting the hash of this key be any easier for a user of this API? Wouldn't that be even harder? Is there some scenario where a user would already have the hash and then would want to use this endpoint?

Note that adding a new version of `get_data_with_proof` that takes in the `TriePath` would also require some significant changes to the `MemoryBackingStore`, since it currently directly uses the key to access the database.

Reply:
> Why would getting the hash of this key be any easier for a user of this API? Wouldn't that be even harder? Is there some scenario where a user would already have the hash and then would want to use this endpoint?

For the client, I think both approaches are equivalently difficult.  The hard part either way is constructing the key.  The difference between sending the key versus the hash of the key is trivial.

However, when comparing the difficulty of implementing the API (including changes to the Clarity DB) to only accept a hash versus the whole metadata key, the former is actually much easier.  First, the Clarity DB already performs lookups into the MARF by hash (albeit via a few layers of indirection), so it's not a very big lift to add a trait method for loading a serialized value by `TrieHash` and adding a MARF method to lookup by `TrieHash`.  Second, by accepting a key, we would commit the API to validating that key so as to property return HTTP 400 responses [1] if the key was semantically invalid.  Adding all of the validation logic and test coverage for it would vastly exceed the work required to support lookups by `TrieHash`.

[1] I'm not a fan of returning a blanket 404 for invalid input.  A 404 specifically means that the resource could not be found, and doesn't tell the user what the actually need to know.  By contrast, any 32-byte value is a valid MARF key, so HTTP 400 would only be necessary if the value was not a hex-encoded 32-byte hash.

Reply:
Sure, I understand that argument. Beware of the changes needed to the `MemoryBackingStore` to support lookups by `TrieHash` -- this PR will unfortunately need to grow significantly. This will be more work than we would hope, but that's probably okay, since changes there are not affecting critical code.

Reply:
It's a bit hard to follow but I've some questions based on my understading; 

>  "any 32-byte value is a valid MARF key, so HTTP 400 would only be necessary if the value was not a hex-encoded 32-byte hash"

We would return a 404 if the hash encodes an invalid key right? So the difference for the client seems to be quite the same to me. It's semantically sligthly more correct to return a 404 for a random 32-byte value than it is for an invalid key, but the result for the client is the same.

> "so it's not a very big lift to add a trait method for loading a serialized value by `TrieHash`"

I'm actually struggling to implement it in the MemoryBackingStore (around [here](https://github.com/stacks-network/stacks-core/blob/24566459e5a8bdb399e0fe2efbfa0dadc571c851/clarity/src/vm/database/sqlite.rs#L323-L329)), because I don't have access `TrieHash::from_data` here.
I don't see how to get from the key to the Hash.
Maybe it's trivial but I just don't know the repository good enough.

Reply:
> We would return a 404 if the hash encodes an invalid key right?

If this API endpoint only took a hash, then the only application responses to expect are 200 (with the serialized data) or 404 (if the hash does not map to a value).  The question of whether or not the hash corresponds to a semantically-valid but absent key or a semantically-invalid key does not need to be answered by the API endpoint.

> It's semantically sligthly more correct to return a 404 for a random 32-byte value than it is for an invalid key, but the result for the client is the same.

That's the thing -- I think it's plausible that this makes a significant difference to the client.  If the client is an application that loads Clarity data, then the difference between receiving a 400 versus a 404 is that the former requires a bugfix (the client made an invalid key) and would likely result in the user being presented with an error condition to report, whereas the latter indicates that the submitted key is valid but not mapped.  Treating both as 404's would lead to a poorer user experience, since users would receive false negatives -- a bug in the application's key construction would be treated the same as the user submitting a query for nonexistent data.

Reply:
I'm happy to add the requisite Clarity DB patches to this PR, once I'm done helping @obycode with https://github.com/stacks-network/stacks-core/pull/5420.

Reply:
I went ahead and added the requisite ClarityDB functionality, and plumbed it through into this RPC endpoint.

Reply:
Thanks a lot @jcnelson, this is super super helpful.
I'm currenlty working on addressing your other remaining comments.
Marking this one a resolved